documentation/design_proposal/002_grafana_v1beta1.md

```diff
@@ -0,0 +1,191 @@
+ DashboardNamespaceSelector *metav1.LabelSelector    `json:"dashboardNamespaceSelector,omitempty"`
+ Jsonnet                    *JsonnetConfig           `json:"jsonnet,omitempty"`
+ GrafanaContainer           *GrafanaContainer        `json:"grafanaContainer,omitempty"`
+}
+
+// GrafanaContainer provides a means to configure the grafana container
+type GrafanaContainer struct {
+ BaseImage         string                   `json:"baseImage,omitempty"`
+ Resources         *v1.ResourceRequirements `json:"resources,omitempty"`
+ ReadinessProbe    *v1.Probe                `json:"readinessProbe,omitempty"`
+ LivenessProbeSpec *v1.Probe                `json:"livenessProbe,omitempty"`
+}
+```
+
+Instead of setting `grafana.spec.baseImage` you would have to do `grafana.spec.grafanaContainer.baseImage`.
+
+Checkout the [grafana_types](002_grafana_types.go) file for a better example. Just overwrite `api/integreatly/v1alpha1/grafana_types.go` with this file.
+
+### Change in defaults
+
+#### Default Security Context
+
+Today we have a number of defaults that would be good to apply to the operator it self and the grafana deployment
+to make it as secure as possible out of the box. One of the problems with this in openshift [SCC](https://docs.openshift.com/container-platform/4.9/authentication/managing-security-context-constraints.html)
+and we have to add some check to see if we are deploying in a openshift environment or not.
+
+One of the configs we should asspire to is something like this in our container seucirty context.
+Openshift don't support providing `runAsGroup` and `runAsUser`, `runAsNonRoot`might work since it gets user and group from openshift by default.
+
+> These settings should also apply to our plugin initContainer.
+
+```.yaml
+    securityContext:
+      allowPrivilegeEscalation: false
+      capabilities:
+        drop:
+        - NET_RAW
+      privileged: false
+      readOnlyRootFilesystem: true
+      runAsGroup: 1000
+      runAsNonRoot: true
+      runAsUser: 1000
+```
+
+These settings also forces us to add some default volumes else grafana won't start. TODO add the volumes needed.
+
+#### Default routes/ingress
+
+Today it's not possible to define none https route. By seperating routes and ingress in to seperate configs in the crd we should be able to seperate these defaults easier.
+In openshift a good default is TLS enabled by default. While in ingress it's not as easy.
+
+So what needs to change?
+
+The default setting for routes should be [edge](https://docs.openshift.com/container-platform/4.9/networking/routes/secured-routes.html) `TLS enabled=true` but enable the possability to run in http mode only.
+
+The default ingress should keep on being TLS enabled=false
+
+#### Default resource requests
+
+To miniamize the risk of our operator and the grafana deployment to be evicted during high load in your clusters we should provide basic resource reuqests for both the operator and the grafana deployment/init container.
+
```

Comment:
I'm hoping that we can further optimize this, to make it as unlikely as possible. Part of the solution should be createOrUpdate, which massively simplifies the logic, and decreases the number of operations, but I think we could also do with a few fewer requests overall, i.e reconciliation periods every minute or so, this should silence a lot of the HTTPUsage alerts the community has reported, even though we like to be up-to-date as much as possible, I don't think we need to check for new resources every 10 seconds, at worst we risk being out of sync for 59 seconds if we settle on a minute reconcilePeriod

Reply:
My comment here isn't so much about current usage. It's more that we should provide a good base to stand on for our users so they don't have to think about resource requests/limits out of the box.

But as you point out the operator it self will most likely decrease it's resource consumption with the new reconcile loop :)